New nodejs binding using the wallet message system and latest neon version.

melatron opened the pull request:
# Description of change

Nodejs binding implementation using the latest neon version and its newest "napi-6". Also using the wallet.rs message system. Everything is similar to how we use the binding in Firefly.


## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Tried all the examples that were written for the old nodejs binding.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

melatron committed:
Initial commit of node binding using message system and latest neon version.

melatron committed:
Initial commit of node binding using message system and latest neon version.

melatron committed:
added gitignore

melatron committed:
added gitignore

melatron committed:
Added different class for AccountManager

melatron committed:
Added different class for AccountManager

melatron committed:
Initial commit for the old API transformed using latest neon version.

melatron committed:
Initial commit for the old API transformed using latest neon version.

melatron committed:
Fixed generate_mnemonic

melatron committed:
Fixed generate_mnemonic

Thoralf-M committed:
Merge branch 'dev' into feat/refactoring-node-binding

melatron committed:
fixed account creation.

melatron committed:
fixed account creation.

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
async fix

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
fixed sync

melatron committed:
added check for undefined.

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
generate address fix

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
added mutexes for AccountManager properties.

melatron committed:
Synced account

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
fixed sync error

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

Thoralf-M committed:
add getNodeInfo, fmt

Thoralf-M committed:
update getnodeinfo

Thoralf-M committed:
fix getNodeInfo

Thoralf-M committed:
handle optional values better, clippy

Thoralf-M committed:
Merge branch 'dev' into feat/refactoring-node-binding

melatron committed:
getNodeInfo async

melatron committed:
apply this.account

melatron committed:
refactoring

melatron committed:
added sync, remove account and refactored.

melatron committed:
list messages and accountManager sync added

melatron committed:
list addresses addet to account

melatron committed:
Added few more methods for the account_manager.

Thoralf-M committed:
enable latest_address, JSON.parse results of generate-, latest- list-Address and listMessages

Thoralf-M committed:
Merge branch 'dev' into feat/refactoring-node-binding

Thoralf-M committed:
fmt

Thoralf-M committed:
clippy

melatron committed:
account manager finished.

melatron committed:
Merge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
More account methods added

melatron committed:
added get_unused_address

melatron committed:
Account finished

melatron committed:
small refactoring

melatron committed:
clippy

Thoralf-M committed:
add alias function and fix balance function

melatron committed:
Fixing examples

melatron committed:
merge

Thoralf-M committed:
fix RemainderValueStrategy

melatron committed:
more examples fix

melatron committed:
removed migration

melatron committed:
env added

Thoralf-M committed:
move dotenv to devDependencies

laumair commented on bindings/nodejs-async/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+    "outFile": "./index.js",
+  },
+  "include": ["lib/**/*"]
+  }
```
Missing new line

laumair commented on bindings/nodejs-async/examples/binding_example/1-create-account.js:
```diff
@@ -0,0 +1,60 @@
+    let account;
+    try {
+      account = manager.getAccount("Alice")
+    } catch (e) { }
```
Probably best to at least log the error here.

laumair commented:
@melatron I think it is best to remove (unncecessary) logs from examples.

laumair commented on bindings/nodejs-async/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+hex = "0.4"
+url = "2.2"
+bee-message = "0.1.1"
+log = "0.4"
```
Missing new line

laumair commented on bindings/nodejs-async/examples/binding_example/1-create-account.js:
```diff
@@ -0,0 +1,60 @@
+
+
+
+// require('dotenv').config()
```
Is this unnecessary?

laumair commented on bindings/nodejs-async/examples/binding_example/2-generate-address.js:
```diff
@@ -0,0 +1,36 @@
+    console.log("Addresses:", addresses)
+}
+
+run()
```
New line missing

laumair replied:
New line missing

laumair commented on bindings/nodejs-async/examples/message_examples/5-backup.js:
```diff
@@ -0,0 +1,22 @@
+    }
+}
+
+run()
```
Missing new line

laumair commented on bindings/nodejs-async/lib/binding/account.js:
```diff
@@ -0,0 +1,132 @@
+const utils = require('../utils.js');
+const { SyncedAccount } = require('./synced_account.js');
+
+let {
```
Not sure we are reassigning these, so should preferably be using `cosnt`

laumair commented:
I would suggest using [JSDoc](https://jsdoc.app/tags-example.html) for all methods and classes. 

[iota.js](https://github.com/iotaledger/iota.js/tree/legacy/iota.js/packages/core/src) has some very good references for it.

laumair commented on bindings/nodejs-async/lib/messages/account.js:
```diff
@@ -0,0 +1,104 @@
+  }
+  async sync(options) {
+    return JSON.parse(await this.messageHandler.sendMessage({
+      id: "1",
```
Imo, we should remove the ids. They are not needed atm.

laumair commented on bindings/nodejs-async/lib/typings/index.d.ts:
```diff
@@ -0,0 +1,355 @@
+// export declare enum MessageType {
```
Why did we comment them out?

laumair commented on bindings/nodejs-async/lib/utils.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+module.exports.promisify = promisify;
+module.exports.RemainderValueStrategy = new RemainderValueStrategy();
```
Missing new line

melatron committed:
refactoring

melatron committed:
Merge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

Thoralf-M committed:
Change error handling in first example

Thoralf-M committed:
add SignerType

melatron committed:
EventListener class fix.

melatron committed:
erge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
events example added

melatron committed:
event example

Thoralf-M committed:
change initLogger to accept an object

melatron committed:
renamed folder

melatron committed:
removed main.js

melatron committed:
refactoring

Thoralf-M committed:
update send example

melatron committed:
removed logs and added finalize impl for eventListener.

melatron committed:
removed ids for messages, refactor a bit and added finalize for accountManager

Thoralf-M committed:
add manager to addEventListener and add removeEventListeners, build in release

Thoralf-M committed:
remove removeEventListeners from example

melatron committed:
added listen and removeEventListeners in AccountManager.

melatron committed:
merged dev

melatron committed:
fmt

Thoralf-M committed:
update workflows

Thoralf-M committed:
add license header

Thoralf-M committed:
fix examples path

Thoralf-M commented on bindings/nodejs/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(clippy::unnecessary_wraps)]
```
Is there a reason to leave this here?

Thoralf-M committed:
update copyright year

Thoralf-M commented on bindings/nodejs/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+
+[dependencies]
+iota-wallet = { path = "../../", version = "0.2.0", features = ["stronghold"] }
+iota-client = { git = "https://github.com/iotaledger/iota.rs", rev = "57c5a33d83be4b8286a54156e9f0d2ac4345e9b9", features = ["mqtt"] }
```
That's only used for the logger I think, would be better to only import bee-common then

Thoralf-M commented on bindings/nodejs/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+backtrace = "0.3"
+hex = "0.4"
+url = "2.2"
+bee-message = "0.1.1"
```
I think we should only use 0.1

Thoralf-M replied:
```suggestion
bee-message = "0.1"
```

Thoralf-M commented on bindings/nodejs/README.md:
````diff
@@ -1,22 +1,107 @@
+```sh
+$ npm install
+$ node
+> require('.').hello()
````
This should be removed?

Thoralf-M commented on bindings/nodejs/lib/binding/syncedAccount.js:
```diff
@@ -0,0 +1,15 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+const addon = require('../../index.node');
```
Is this import needed here?

Thoralf-M commented on bindings/nodejs/lib/index.js:
```diff
@@ -1,143 +1,52 @@
 
-Account.prototype.sync = promisify(Account.prototype.sync)
-Account.prototype.isLatestAddressUnused = promisify(Account.prototype.isLatestAddressUnused)
+// todo remove this
```
needs to be removed

Thoralf-M commented on bindings/nodejs/lib/index.js:
```diff
@@ -1,143 +1,52 @@
-const addon = require('../build/Release')
-let { AccountManager, Account, EventListener, initLogger } = addon
+const addon = require('../index.node');
+// const mh = require("./messages/messageHandler.js");
```
Can we export both things at the same time so one doesn't need to uncomment one?

Thoralf-M commented on bindings/nodejs/lib/messages/accountManager.js:
```diff
@@ -0,0 +1,75 @@
+let { MessageHandler } = mh
+let { Account } = acc;
+
+const _id = "1";
```
Unused variable, can be removed?

Thoralf-M commented on bindings/nodejs/lib/messages/messageHandler.js:
```diff
@@ -0,0 +1,23 @@
+const addon = require('../../index.node');
+const utils = require('../utils.js');
+
+console.log(utils);
```
```suggestion
```

Thoralf-M commented on bindings/nodejs/src/classes/message_handler.rs:
```diff
@@ -0,0 +1,164 @@
+    Error,
+};
+
+const DUMMY_ID: &str = "1";
```
Please add a comment why this is needed

Thoralf-M commented on bindings/nodejs/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+        .map(char::from)
+        .take(10)
+        .collect();
+    map.insert(id.clone(), Arc::new(RwLock::new(synced_account)));
```
Shouldn't we check if the id is already used? I know the probability for this to happen is very low, but not impossible

Thoralf-M commented on bindings/nodejs/src/types.rs:
```diff
@@ -157,3 +148,41 @@ impl From<ClientOptionsDto> for ClientOptions {
+    pub persist_events: bool,
+    #[serde(rename = "allowCreateMultipleEmptyAccounts", default)]
+    pub allow_create_multiple_empty_accounts: bool,
+
```
```suggestion
```

Thoralf-M commented on bindings/nodejs/test-database/db/CURRENT:
```diff
@@ -0,0 +1 @@
+MANIFEST-000003
```
The whole test-database folder should be deleted?

Thoralf-M commented on src/account/mod.rs:
```diff
@@ -53,6 +53,16 @@ pub enum AccountIdentifier {
     Index(usize),
 }
 
+impl<'de> Deserialize<'de> for AccountIdentifier {
```
Why didn't it work with the derived Deserialize?

Thoralf-M replied:
"it won't work, because it won't Deserialize the correct enum value"

Thoralf-M commented on bindings/nodejs/examples/binding_examples/2-generate-address.js:
```diff
@@ -1,33 +1,36 @@
 
     const account = manager.getAccount('Alice')
-    console.log('Account:', account.alias())
+    // console.log('Account:', account.alias())
```
```suggestion
    console.log('Account:', account.alias())
```

Thoralf-M committed:
update node url

melatron committed:
review fixes

Thoralf-M commented on bindings/nodejs/lib/binding/syncedAccount.js:
```diff
@@ -0,0 +1,13 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+const addon = require('../../index.node');
```
```suggestion
```

Thoralf-M committed:
remove unused import

Thoralf-M commented on bindings/nodejs/examples/binding_examples/7-events.js:
```diff
@@ -5,11 +5,12 @@
         storagePath: './alice-database'
     })
-
+    console.log("Setting stronghold password.");
```
```suggestion
```

Thoralf-M commented on bindings/nodejs/examples/binding_examples/7-events.js:
```diff
@@ -5,11 +5,12 @@
     })
-
+    console.log("Setting stronghold password.");
+    console.log(process.env.SH_PASSWORD)
```
```suggestion
```

melatron committed:
message examples fix

melatron committed:
Merge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
license template added

melatron committed:
clippy

Thoralf-M committed:
remove empty line

Thoralf-M committed:
uncomment log

Thoralf-M committed:
remove log

Thoralf-M committed:
remove log

Thoralf-M committed:
update dependencies

Thoralf-M commented on bindings/nodejs/src/classes/account/mod.rs:
```diff
@@ -0,0 +1,839 @@
+        let _ = sender.send(index);
+    });
+
+    Ok(cx.number(receiver.recv().unwrap() as f64))
```
Shouldn't this be an u64?

Thoralf-M commented on bindings/nodejs/src/classes/account/mod.rs:
```diff
@@ -0,0 +1,839 @@
+        &&cx.this()
+            .downcast_or_throw::<JsBox<Arc<AccountWrapper>>, FunctionContext>(&mut cx)?,
+    );
+    // let account_wrapper = Arc::clone(&&cx.argument::<JsBox<Arc<AccountWrapper>>>(cx.len() - 2)?);
```
```suggestion
```

Thoralf-M commented on bindings/nodejs/src/classes/account/mod.rs:
```diff
@@ -0,0 +1,839 @@
+        let _ = sender.send(count);
+    });
+
+    Ok(cx.number(receiver.recv().unwrap() as f64))
```
Same here, shouldn't this be an u64?

Thoralf-M commented on bindings/nodejs/src/classes/account/mod.rs:
```diff
@@ -0,0 +1,839 @@
+            .downcast_or_throw::<JsBox<Arc<AccountWrapper>>, FunctionContext>(&mut cx)?,
+    );
+    let message_id =
+        MessageId::from_str(cx.argument::<JsString>(0)?.value(&mut cx).as_str()).expect("invalid message id length");
```
```suggestion
        MessageId::from_str(cx.argument::<JsString>(0)?.value(&mut cx).as_str()).expect("invalid message id");
```

Thoralf-M committed:
change error message

melatron committed:
readme

melatron committed:
Merge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

melatron committed:
disabled needless_borrow

Thoralf-M committed:
update installation part

melatron committed:
fmt

melatron committed:
Merge branch 'feat/refactoring-node-binding' of github.com:iotaledger/wallet.rs into feat/refactoring-node-binding

Thoralf-M committed:
update examples paths

Thoralf-M committed:
update testurl

Thoralf-M committed:
remove migration example

Thoralf-M committed:
change entrypoint

Thoralf-M committed:
fix setClientOptions

Thoralf-M committed:
add index method and parse message

Thoralf-M committed:
export RemainderValueStrategy from the main module

Thoralf-M committed:
fix listMessages

Thoralf-M committed:
expose messageCount

Thoralf-M committed:
Add OutputKind to allow dustAllowanceOutputs creation

melatron committed:
added stop and start background sync methods to AccountManager

Thoralf-M committed:
Merge branch 'dev' into feat/refactoring-node-binding

Thoralf-M committed:
add comment about urls

Thoralf-M committed:
Merge branch 'feat/refactoring-node-binding' of https://github.com/iotaledger/wallet.rs into feat/refactoring-node-binding

Thoralf-M committed:
update import path

melatron committed:
typo fix

melatron committed:
messages fix

Thoralf-M committed:
update package name

Commit:
chore: integrate eslint & prettier (#744)

Thoralf-M committed:
log object

Commit:
chore: do minor formatting changes

Commit:
chore: rename variable

Commit:
chore: remove hardcoded password

Commit:
chore: properly log object

Commit:
chore: remove unnecessary log

Commit:
chore: rename variable

Commit:
hotfix: fix typo

Commit:
chore: do some clean up

- Rename some logs;
- Rename example files for consistency;
- Require dotenv in example files

Commit:
chore: do minor formatting changes

Commit:
chore: fix formatting

Commit:
docs: update README

Commit:
chore: remove test db

Commit:
docs: update README

Commit:
docs: update refs to example in docs file

Thoralf-M committed:
add event listener to account manager and add setClientOptions (#745)

* add event listener to account manager and add setClientOptions

* lint

* clean example

Commit:
chore: use tabWidth = 4

Thoralf-M committed:
typo, order exports

rajivshah3 commented on bindings/nodejs/package.json:
```diff
@@ -1,29 +1,33 @@
+        "test": "cargo test"
+    },
+    "author": "Antoni Dikov <antoni.dikov@iota.org>",
+    "license": "MIT",
```
```suggestion
    "license": "Apache-2.0",
```
Just realized it said `MIT` before this PR, but it should match the license of `wallet.rs`

rajivshah3 commented on bindings/nodejs/package.json:
```diff
@@ -1,29 +1,33 @@
-    "build:docs": "./node_modules/.bin/jsdoc lib/index.js"
-  },
-  "devDependencies": {
-    "electron-build-env": "^0.2.0",
-    "jsdoc": "^3.6.6",
-    "prebuild": "^10.0.1"
```
Are we no longer providing prebuilt binaries?

Commit:
Update bindings/nodejs/package.json

Co-authored-by: Rajiv Shah <rajivshah1@icloud.com>

melatron committed:
added prebuild-install

Thoralf-M committed:
add more debug logs

Thoralf-M committed:
add changelog
